src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs

```diff
@@ -5375,6 +5375,104 @@ .maxstack  3
   IL_004f:  box        ""int""
   IL_0054:  stfld      ""object C.<>c__DisplayClass0_0.o""
   IL_0059:  ret
+}");
+        }
+
+        [WorkItem(44720, "https://github.com/dotnet/roslyn/issues/44720")]
+        [Fact]
+        public void LambdaInsideGenericLocalFunctionInsideLoop_CompilesCorrectly()
+        {
+            string source =
+                @"using System;
+using System.Collections.Generic;
+
+static class Program
+{
+    private static void Main()
+    {
+        TestMethod(string.Empty);
+    }
+
+    private static void TestMethod<T>(T param)
+    {
+        var message = string.Empty;
+
+        for (int i = 0; i < 1; i++)
```

Comment:
>for (int i = 0; i < 1; i++) [](start = 8, length = 27)

Is there an affected scenario that doesn't involve a loop?

Reply:
According to [the condition (shouldCacheForStaticMethod || shouldCacheInLoop)](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs#L1586) there should be a scenario involving a static method instead of a loop. Haven't tried though.

Reply:
No, it seems to be reproducable only using a loop. Roslyn does not cache the delegate for static methods if the method is generic (wondering if that can be changed with the check I've added).

Reply:
> No, it seems to be reproducable only using a loop. Roslyn does not cache the delegate for static methods if the method is generic (wondering if that can be changed with the check I've added).

Is it possible that ```shouldCacheForStaticMethod``` is always set correctly and only ```shouldCacheInLoop``` is incorrect and we should adjust the place where it is calculated to do the same thing as we do while calculating ```shouldCacheForStaticMethod```?


---
In reply to: [437690642](https://github.com/dotnet/roslyn/pull/44939#discussion_r437690642) [](ancestors = 437690642)

Reply:
I've dug into the logic behind the second part of the condition (caching into a local instead of a field), and it seams to be a dead code.

`F.SynthesizedLocal` is always invoked with `syntax = null` and `kind = CachedAnonymousMethodDelegate` ([code](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs#L1615)) and that leads to failed debug assertion in `SynthesizedLocal..ctor` ([code](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedLocal.cs#L47)) or `NullReferenceException` in `LambdaUtilities.GetDeclaratorPosition` ([code](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs#L479)).

Considering that, the field is the only possible option to do caching, I can remove the dead code and move my check outside the if statement.

Would appreciate if you confirm if that code is a dead code.

Reply:
> Would appreciate if you confirm if that code is a dead code.

At the moment, we are really heads down trying to complete the remaining feature work. One way to confirm that your assumption is incorrect is to make a change to throw when the dead code is reached, run all tests and see if anything breaks. If nothing breaks, there is still a chance that there is a test gap.

Reply:
Have already done that. None of the tests executes that branch. Haven't tried dogfooding the compiler into itself yet though.

I've tried to come up with a test case to cover that code. That's how I concluded that the code just cannot be executed successfully at all. 🙂

Reply:
> Have already done that. None of the tests executes that branch. Haven't tried dogfooding the compiler into itself yet though.

> I've tried to come up with a test case to cover that code. That's how I concluded that the code just cannot be executed successfully at all.

I suggest opening an issue about that and capturing results of your investigation there. I don't think we should do anything about that in context of this PR because it doesn't appear to be related to the issue we are trying to address.

---
In reply to: [445836446](https://github.com/dotnet/roslyn/pull/44939#discussion_r445836446) [](ancestors = 445836446)